Add CoreCLR support for android GC bridge

BrzVlad opened the pull request:
This change adds runtime support for the GCBridge api described in https://github.com/dotnet/runtime/issues/115506 and to be used on android. It includes most of initial work from https://github.com/dotnet/runtime/pull/114184.

When the GCBridge feature is used, at the start of the application `JavaMarshal.Initialize` is called. This will provide to the runtime a native callback (`markCrossReferences`) to be called during the GC when the collection takes place. During GC, we compute the set of strongly connected components containing bridge objects that are dead in the .NET space. These SCCs are passed to the callback so the .NET android implementation would reproduce the links between the java counterparts in order to determine whether the .NET object needs to be collected or not (The constraint is that the C# peer keeps the Java Peer alive and vice verse. We make no effort to handle finalization, so a resurrected C# object can have the Java Peer collected). Once the .NET Android runtime does the java collection it will report back to the runtime with the list of bridge objects that can be freed and with the previously passed SCC related pointers to be freed.

A bridge object is an object that can have a JavaPeer. The CoreCLR runtime has no insight into this, the only thing it understands are cross reference handles. These are GCHandles that have an additional pointer associated with them, so additional information related to the java peer can be attached. Objects that have a cross reference handle allocated, will always survive the current GC collection, because we can't collect them until we get permission from the Java world. Once the cross reference gchandle is freed, the associated object becomes ordinary, detached from any java peer, and it is free for collection in the .NET heap.

At the end of mark phase, during GC, we iterate over all cross reference handles. When we encounter a handle with target that hasn't yet been marked, we add it to a list (these objects will have to be marked so they remain alive after this collection, given we need to probe the java world first). Once we obtained the set of dead bridge objects, we apply the tarjan algorithm (this algorithm is ported directly from mono's implementation). This algorithm will operate on the dead object graph, reachable from the initial set of dead bridge objects. In order to implement this secondary scanning mechanism, for objects that we reach, we hijack the object header with a `ScanData` that contains all information relevant to the SCC building algorithm. Once we finished building the SCCs, while still in the GC, we callback into the .NET Android via `TriggerClientBridgeProcessing` that will end up calling the mark cross reference callback provided by `JavaMarshal.Initialize`. This callback will have to dispatch the neceesary work for another thread to run, since it needs to return quickly, for the C# GC to continue its execution.

Because the world gets resumed without having decided yet whether the bridge objects will be alive or dead, for weak references, we would need to wait for the java bridge processing to finish before we can resolve the Target. Aside from the general problem of resurrecting a C# peer that has the Java Peer collected, this mechanism will be used internally by the .NET Android in order to manually manage liveness of these bridge objects, in the scenario of calling Dispose on an object. This synchronisation will be used at the core of .NET Android Runtime interop. In order to implement this, weak refs for bridge objects are not nulled during GC (these objects are promoted during collection) but rather at the finishing stage of bridge processing. This change is conservative and adds bridge waiting only for WeakReference, not when using GCHandle, following the existing approach in COM. 

This PR adds a few tests in the runtime tests. The tests have a native counterpart that acts as the client bridge, not doing anything, just doing random sleep instead of doing the Java collection. The test creates a set of objects with certain links between then, creates weak refs to the BridgeObjects and then doesn't reference anything else. Depending on the built graph, it expects a certain number of SCCs and cross refs constructed by the tarjan algorithm, and then reports all bridge objects as alive or dead. The test will also check to see if the Target for all the weak refs is the expected one.

The gcbridge doesn't consume much memory. A collection for a heavier app can end with hundreds of SCCs and xrefs. For such a scenario, the gcbridge is expected to consume hundreds of KBs. Most of this memory is represented by data for ScanData, ColorData and stacks used by tarjan algorithm. These data structures have their capacity increased when necessary, so for most collections there is no new memory allocation, the existing storage is reused. For a few other data structures, like xrefs arrays and data allocated to be passed to the bridge client, new allocations from scratch happen at each collection. While the gc bridge can end up consuming hundreds of KB for heavy scenarios, maybe a few MB in extreme theoretical cases, less than 10% of this memory is expected to be allocated during collection, the rest should be reused.

BrzVlad committed:
Add support for the new cross reference GCHandle type

From Aaron's implementation.

BrzVlad committed:
Add interop glue for bridge

From Aaron's implementation

BrzVlad committed:
Add GCBridge test

BrzVlad committed:
Implement scanning of bridge objects and construction of SCCs

from Maoni

BrzVlad committed:
Add tarjan bridge computation

BrzVlad committed:
Fix object validation

Checking if the object is promoted was validating the next object header in debug builds. During bridge tarjan computation, we patch the object header for some objects in order to store data used by the bridge algorithm, so we need to disable this validation.

BrzVlad committed:
bgc and multiple heap support for bridge obj processing

BrzVlad committed:
Add incorrect waiting for gc bridge to finish

BrzVlad committed:
Update code to new API

BrzVlad committed:
Clear weakrefs for bridge objects with collected java peers

BrzVlad committed:
Wait for bridge processing while resolving the gchandle to prevent races

BrzVlad committed:
Maoni review

HANDLE_MAX_INTERNAL_TYPES value

new instead of malloc

assert for allocation failure

Reuse memory for ColorData and ScanData between collections. We still do alloc/free for other type of data, for example for arrays representing edges between SCCs.

Actually print class name when enabling tarjan bridge logs.

Add separate IsPromoted method to the gc interface

Rename TriggerGCBridge to TriggerClientBridgeProcessing to be more specific about what it is doing.

BrzVlad committed:
Aaron review

BrzVlad committed:
Update to new API

jkotas commented on src/coreclr/vm/interoplibinterface_shared.cpp:
```diff
@@ -154,3 +154,142 @@ void Interop::OnAfterGCScanRoots(_In_ bool isConcurrent)
+
+namespace
+{
+    Volatile<BOOL> g_GCBridgeActive = FALSE;
```
```suggestion
    Volatile<bool> g_GCBridgeActive = false;
```
Nit: It is fine to use native C/C++ bool for internal details.

BrzVlad committed:
Fix build

Include classes that are part of the Java Marshal API on mono as well

Disable test on mono and nativeaot.

jkotas commented on src/coreclr/vm/interoplibinterface_shared.cpp:
```diff
@@ -154,3 +154,142 @@ void Interop::OnAfterGCScanRoots(_In_ bool isConcurrent)
+
+bool Interop::IsGCBridgeActive()
+{
+    CONTRACTL
+    {
+        MODE_COOPERATIVE;
+    }
+    CONTRACTL_END;
```
```suggestion
    LIMITED_METHOD_CONTRACT;
```
This should fix the CI failure

jkoritzinsky commented on src/coreclr/vm/interoplibinterface_shared.cpp:
```diff
@@ -154,3 +154,138 @@ void Interop::OnAfterGCScanRoots(_In_ bool isConcurrent)
+    _In_ size_t ccrsLen,
+    _In_ ComponentCrossReference* ccrs)
+{
+    CONTRACTL
+    {
+        NOTHROW;
+        GC_NOTRIGGER;
+    }
+    CONTRACTL_END;
```
Can we add an a `PRECONDITION` for this being called during GC suspension? Makes it easier to understand the flow without tracing all of the callers (same for JavaNative::TriggerClientBridgeProcessing) especially since the contract doesn't have a GC mode requirement.

jkoritzinsky commented on src/tests/Interop/GCBridge/BridgeTest.cs:
Can we change the tests here to use Asserts and not the legacy "return 100 == success" model?

BrzVlad commented on src/coreclr/vm/interoplibinterface_shared.cpp:
```diff
@@ -154,3 +154,142 @@ void Interop::OnAfterGCScanRoots(_In_ bool isConcurrent)
+    // This runs during GC while the world is stopped, no synchronisation required
+    if (!g_bridgeFinished)
+    {
+        g_bridgeFinished = new CLREvent();
+        g_bridgeFinished->CreateManualEvent(false);
```
I'm not sure why this should be associated with java code. Seems to me that even if we were to interop with another garbage collected language via gcbridge, we would still need to do this weak reference waiting for bridge to finish. So I think that `JavaMarshal` is just about the exposed C# api. If we were to add another api for another language, wouldn't it have the exact same format and call the exact same methods in `Interop::*`. Why should we move stuff out of the shared location ?

Having said that, the distinction between `FEATURE_JAVAMARSHAL` and `FEATURE_GCBRIDGE` is quite minimal and I feel it just adds confusion currently. In my humble opinion, I would just have everything under `FEATURE_JAVAMARSHAL` instead. If we are to interop with another language in the future, we can easily make the split then, I don't feel like we need it now.

BrzVlad committed:
Use cross refs args directly in more places

BrzVlad committed:
Change unreachable handle length from int to size_t

BrzVlad committed:
Remove FEATURE_GCBRIDGE

BrzVlad committed:
Move interop javamarshal code out of interop shared file

Create reset event during javamarshal initialize

BrzVlad committed:
Use asserts in test

BrzVlad committed:
Add bridge waiting to WeakReference<T> as well

jkotas commented on src/coreclr/vm/interoplibinterface_java.cpp:
```diff
@@ -0,0 +1,236 @@
+    _ASSERTE(g_GCBridgeActive);
+
+    // Mark the GCBridge as inactive.
+    // This much be synchronized with the GC so switch to cooperative mode.
```
```suggestion
    // This must be synchronized with the GC so switch to cooperative mode.
```

jkotas commented on src/coreclr/vm/interoplibinterface_java.cpp:
```diff
@@ -0,0 +1,236 @@
+        for (size_t i = 0; i < length; i++)
+            pHandleManager->DestroyHandleOfUnknownType(((OBJECTHANDLE*)unreachableObjectHandles)[i]);
+
+        g_GCBridgeActive = FALSE;
```
```suggestion
        g_GCBridgeActive = false;
```
Nit

BrzVlad replied:
This was supposed to be `BOOL` but seems like I mixed the code. My understanding is that `BOOL` is `int32` while `bool` could be `int8` and we use `InterlockedCompareExchange` which doesn't support types smaller than 4 bytes, therefore we would have to use `BOOL`.

Having said that, using CAS on initialize seems kind of redundant to me. I think calling `JavaMarshal.Initialize` multiple times should be undefined behavior. @jkotas @AaronRobinsonMSFT Are you fine with removing the CAS during initialize ?

jkotas replied:
I think you are confusing this with `g_Initialized`. `g_GCBridgeActive` is `bool` and other places use lower case true/false, so this comment is just a small consistency nit.

For `g_Initialized`, I do not think you really need it. You can do InterlockedCompareExchange on `g_MarkCrossReferences` to ensure that it gets initialized only once.

AaronRobinsonMSFT replied:
>  I think calling JavaMarshal.Initialize multiple times should be undefined behavior.

It shouldn't be undefined, it is very defined and illegal. The prototype will throw if it is already initialized. This is the same approach for ObjectiveC.

jkotas commented on src/tests/Interop/GCBridge/BridgeTest.cs:
```diff
@@ -0,0 +1,275 @@
+using System;
```
All C# and C++ source files should have license header (incl tests)

BrzVlad committed:
Review

jkotas commented on src/coreclr/gc/env/gctoeeinterface.standalone.inl:
```diff
@@ -49,6 +49,11 @@ namespace standalone
             return ::GCToEEInterface::RefCountedHandleCallbacks(pObject);
         }
 
+        void TriggerClientBridgeProcessing(MarkCrossReferencesArgs* args)
```
I am worried about the existing standalone GC scenarios that the GC team cares about being broken. As implemented in the PR currently, I do not think you can use new GC with old runtime, and vice versa.

jkotas commented on src/coreclr/gc/env/gcenv.ee.h:
```diff
@@ -98,6 +100,8 @@ class GCToEEInterface
 
     static uint64_t GetThreadOSThreadId(Thread* thread);
+
+    static const char* GetMethodTableDebugName(MethodTable* pMT);
```
The method signature needs to be different then. We do not keep around fully populated type name in release.

Maoni0 replied:
> The method signature needs to be different then. 

yep, should be different. 

a partially populated type name should be sufficient for this purpose, IMO.

BrzVlad replied:
The reason I added this method is because I don't understand how to obtain a method table friendly name to print. I don't see how `FireGCAllocationTick_V4` approach could help, since this method is implemented in the VM so it can use runtime APIs that are not accessible in the GC, like `TypeHandle`.

From what I can tell, stress log is used for some key GC events, even during Release, while in my case the logging was more like full on debugging of every step done by the algorithm (which is currently enabled only on custom build of runtime), so not sure if it fits ideally in the stress log use case with the current level of logging. So, assuming I'm not using stress log, any other way to print method table friendly name ? Is it ok if I keep this API as is for now ? I think in the future it would be useful to have some sort of separate stresslog logging so we have better ways to investigate issues in Release builds, but I'd rather not delay and add too much stuff on top of this PR.

jkoritzinsky replied:
I believe there's the dprintf style of logging in the GC. That can be backed by the stresslog, so for those you can use the special mechanisms to output a MethodTable into the log.

BrzVlad replied:
I decided to give up on this API in the end, since I'm not convinced at this moment if it makes sense to have it. I'll revisit this later to use stresslog instead, so we can have logging during release builds as well.

AaronRobinsonMSFT commented on src/tests/Interop/CMakeLists.txt:
```diff
@@ -99,6 +99,7 @@ endif(CLR_CMAKE_TARGET_WIN32)
 if(CLR_CMAKE_TARGET_UNIX)
     add_subdirectory(COM/ComWrappers/MockReferenceTrackerRuntime)
     add_subdirectory(COM/ComWrappers/WeakReference)
+    add_subdirectory(GCBridge)
 endif(CLR_CMAKE_TARGET_UNIX)
```
Remove the `if(CLR_CMAKE_TARGET_UNIX)` itself. The three `add_subdirectory()` can be moved to line 66. Remove the two lines at 86 and 86 that are only for Windows. All three build everywhere.

BrzVlad committed:
review

BrzVlad committed:
High precision timestamp

BrzVlad committed:
Naming consistency in gcbridge.cpp, matching rest of the runtime

BrzVlad committed:
Fix interface breakage

BrzVlad commented on src/tests/Interop/GCBridge/BridgeTest.cs:
```diff
@@ -0,0 +1,277 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+using System;
```
Seems like we are not running any runtime tests with release runtime, otherwise it would be crashing. Not sure of an easy way to do it. I guess I could add a feature switch that the test checks, following the example of `IsBuiltInComSupported` ?

jkotas replied:
> Seems like we are not running any runtime tests with release runtime

I think we do - they are just not part of the default CI set.

jkotas commented:
/azp list

jkotas commented:
/azp run runtime-coreclr outerloop

jkotas commented:
/azp run runtime-nativeaot-outerloop

hoyosjs commented:
/azp run coreclr-release-outerloop-nightly

jkotas commented:
The new test is failing in coreclr-release-outerloop-nightly. It should be fixed (e.g. you can catch PNSE exception during initialization when not on Android and skip the rest of the test).

BrzVlad committed:
Fix test/JavaMarshal on platforms where it is not enabled

The test used to fail and the JavaMarshal api would crash due to missing internal calls.

jkotas commented:
/azp run coreclr-release-outerloop-nightly

BrzVlad committed:
Share GetHighPrecisionTimeStamp across GC code

Maoni0 commented on src/coreclr/gc/gcinterface.h:
```diff
@@ -11,11 +11,11 @@
 // to bump the minor version number. GCs and EEs with minor version number
 // mismatches can still interoperate correctly, with some care.
-#define GC_INTERFACE_MINOR_VERSION 4
+#define GC_INTERFACE_MINOR_VERSION 5
```
I don't think we actually use the minor version :) nowhere do we use this version for anything (we just log a message if the version in the standalone GC is lower). the original intention for this was that it could be useful to detect some non breaking behavior and do something different but I don't think we've since made use of that.

vitek-karas commented:
@Maoni0 @jkotas - is this ready? Could it be approved and merged?

jkotas reviewed (approved):
It would be best for @Maoni0 and @AaronRobinsonMSFT to sign-off on this one, but they are oof currently. If there is any additional feedback, it can be incorporated once they are back.

mangod9 commented:
/azp run runtime-coreclr outerloop

mangod9 commented:
I have just triggered an outerloop run given the surface area of this change. if that is looking good we can certainly merge and do subsequent changes subsequently if needed

BrzVlad committed:
Fix mark function used during BGC finishing pause

mangod9 commented:
outerloop is good. Assume browser-wasm are known issues?
